platforms/jvm/code-quality/src/main/java/org/gradle/api/plugins/quality/internal/CodeNarcInvoker.java

```diff
@@ -0,0 +1,153 @@
+                    }
+                });
+        } catch (Exception e) {
+            if (e.getMessage().matches("Exceeded maximum number of priority \\d* violations.*")) {
+                String message = "CodeNarc rule violations were found.";
+
+                // Find all reports that produced a file
+                List<CodeNarcActionParameters.EnabledReport> reportsWithFiles = reports.stream()
+                    .filter(it -> !it.getName().get().equals("console"))
+                    .collect(Collectors.toList());
+                // a report file was generated
+                if (!reportsWithFiles.isEmpty()) {
+                    CodeNarcActionParameters.EnabledReport humanReadableReport = reportsWithFiles.stream().filter(it -> it.getName().get().equals("html"))
+                        .findFirst()
+                        .orElse(null);
+                    if (humanReadableReport == null) {
+                        humanReadableReport = reportsWithFiles.stream().filter(it -> it.getName().get().equals("text"))
+                            .findFirst()
+                            .orElse(null);
+                    }
+                    if (humanReadableReport == null) {
+                        humanReadableReport = reportsWithFiles.stream().filter(it -> it.getName().get().equals("xml"))
+                            .findFirst()
+                            .orElse(null);
+                    }
+                    // Prefer HTML > text > XML and don't include a link if we don't recognize the report format
+                    if (humanReadableReport != null) {
+                        String reportUrl = new ConsoleRenderer().asClickableFileUrl(humanReadableReport.getOutputLocation().getAsFile().get());
+                        message += " See the report at: " + reportUrl;
+                    }
+                }
+
+                if (ignoreFailures) {
+                    LOGGER.warn(message);
+                    return;
+                }
+                throw new MarkedVerificationException(message, e);
+            }
+            if (e.getMessage().contains("codenarc doesn't support the nested \"classpath\" element.")) {
+                String message = "The compilationClasspath property of CodeNarc task can only be non-empty when using CodeNarc 0.27.0 or newer.";
+                throw new GradleException(message, e);
+            }
+            throw e;
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    static void setLifecycleLogLevel(AntBuilderDelegate ant, String lifecycleLogLevel) {
+        try {
+            Object project = ant.getBuilder().getClass().getMethod("getProject").invoke(ant.getBuilder());
+            List<Object> buildListeners = (List<Object>) project.getClass().getMethod("getBuildListeners").invoke(project);
+            for (Object it : buildListeners) {
+                if (it.getClass().getName().equals(AntLoggingAdapter.class.getName())) {
```

Comment:
Can we keep the comment:
```suggestion
                // We cannot use instanceof or getClass()==AntLoggingAdapter since they're in different class loaders
                if (it.getClass().getName().equals(AntLoggingAdapter.class.getName())) {
```